plugins/teststeps/teststeps.go

```diff
@@ -27,53 +29,80 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
-	for {
-		select {
-		case target := <-ch.In:
-			if target == nil {
-				// no more targets incoming
-				return nil
+	type tgtErr struct {
+		target *target.Target
+		err    error
+	}
+	tgtErrCh := make(chan tgtErr)
+	tgtInFlight := int32(0)
+	noMoreTargetsCh := make(chan struct{})
+
+	go func(tgtInFlight *int32) {
+		defer func() {
+			log.Debugf("%s: ForEachTarget: exiting incoming loop, targets inflight %d", pluginName, tgtInFlight)
+		}()
+		for {
+			tgt := <-ch.In
+			log.Debugf("%s: ForEachTarget: received target %s", pluginName, tgt)
+
+			if tgt == nil {
+				log.Debugf("%s: ForEachTarget: all targets have been received", pluginName)
+				noMoreTargetsCh <- struct{}{}
+				return
 			}
-			log.Debugf("%s: ForEachTarget: received target %s", pluginName, target)
-			errCh := make(chan error)
+
 			go func() {
-				log.Debugf("%s: ForEachTarget: calling function on target %s", pluginName, target)
-				errCh <- f(cancel, pause, target)
+				tgtErrCh <- tgtErr{target: tgt, err: f(cancel, pause, tgt)}
 			}()
+			atomic.AddInt32(tgtInFlight, 1)
+		}
+	}(&tgtInFlight)
 
-			select {
-			case err := <-errCh:
-				if err != nil {
-					select {
-					case ch.Err <- cerrors.TargetError{Target: target, Err: err}:
-						log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, target, err)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
-				} else {
-					select {
-					case ch.Out <- target:
-						log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, target)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
+	noMoreTargets := false
+	defer func() {
+		log.Debugf("%s: ForEachTarget: exiting outgoing loop, targets inflight %d, the last target received %v", pluginName, tgtInFlight, noMoreTargets)
+	}()
+	for {
+		select {
+		case te := <-tgtErrCh:
+			atomic.AddInt32(&tgtInFlight, -1)
```

Comment:
Do we really need to keep track of `tgtInFlight`? Since I think we need a wait group, could we just wait for `ch.In` to close, and then wait for the wait group to terminate?

Reply:
Yeah, we could use a wait group here in this case.